### PR TITLE
Fix not being able to open a card from a dashboard with sandboxing

### LINKED ISSUE #29076: Cannot get into a card from a dashboard on "no self-service" permissions
**Describe the bug**
I have a feeling that this issue is very similar to https://github.com/metabase/metabase/issues/28947. If a user has a no self-service permission to a single table then the user will be able to see a card from the dashboard, but it won't be able to see the question when it clicks the title of the card

**Logs**
<details>
```
{:database_id 1,
 :started_at #t "2023-03-09T13:13:08.595658Z[GMT]",
 :error_type :missing-required-permissions,
 :json_query
 {:database 1,
  :query {:source-table 7},
  :type "query",
  :parameters [],
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :native nil,
 :status :failed,
 :class clojure.lang.ExceptionInfo,
 :stacktrace
 ["--> query_processor.middleware.permissions$perms_exception.invokeStatic(permissions.clj:33)"
  "query_processor.middleware.permissions$perms_exception.doInvoke(permissions.clj:27)"
  "query_processor.middleware.permissions$perms_exception.invokeStatic(permissions.clj:30)"
  "query_processor.middleware.permissions$perms_exception.invoke(permissions.clj:27)"
  "query_processor.middleware.permissions$fn__57214$check_ad_hoc_query_perms__57219$fn__57220.invoke(permissions.clj:87)"
  "query_processor.middleware.permissions$fn__57214$check_ad_hoc_query_perms__57219.invoke(permissions.clj:83)"
  "query_processor.middleware.permissions$fn__57246$check_query_permissions_STAR___57251$fn__57252.invoke(permissions.clj:102)"
  "query_processor.middleware.permissions$fn__57246$check_query_permissions_STAR___57251.invoke(permissions.clj:92)"
  "query_processor.middleware.permissions$check_query_permissions$fn__57265.invoke(permissions.clj:111)"
  "metabase_enterprise.advanced_permissions.query_processor.middleware.permissions$check_download_permissions$fn__97936.invoke(permissions.clj:128)"
  "metabase_enterprise.sandbox.query_processor.middleware.column_level_perms_check$maybe_apply_column_level_perms_check$fn__99329.invoke(column_level_perms_check.clj:31)"
  "query_processor.middleware.mbql_to_native$mbql__GT_native$fn__60651.invoke(mbql_to_native.clj:24)"
  "query_processor$fn__63229$combined_post_process__63234$combined_post_process_STAR___63235.invoke(query_processor.clj:243)"
  "query_processor$fn__63229$combined_pre_process__63230$combined_pre_process_STAR___63231.invoke(query_processor.clj:240)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__61653$fn__61658.invoke(resolve_database_and_driver.clj:36)"
  "driver$do_with_driver.invokeStatic(driver.clj:88)"
  "driver$do_with_driver.invoke(driver.clj:84)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__61653.invoke(resolve_database_and_driver.clj:35)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__57578.invoke(fetch_source_query.clj:317)"
  "query_processor.middleware.store$initialize_store$fn__57770$fn__57771.invoke(store.clj:12)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:47)"
  "query_processor.store$do_with_store.invoke(store.clj:41)"
  "query_processor.middleware.store$initialize_store$fn__57770.invoke(store.clj:11)"
  "query_processor.middleware.normalize_query$normalize$fn__61926.invoke(normalize_query.clj:23)"
  "metabase_enterprise.audit_app.query_processor.middleware.handle_audit_queries$handle_internal_queries$fn__68528.invoke(handle_audit_queries.clj:131)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__58874.invoke(constraints.clj:54)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__61865.invoke(process_userland_query.clj:148)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__62237.invoke(catch_exceptions.clj:169)"
  "query_processor.reducible$async_qp$qp_STAR___52680$thunk__52682.invoke(reducible.clj:101)"
  "query_processor.reducible$async_qp$qp_STAR___52680.invoke(reducible.clj:107)"
  "query_processor.reducible$async_qp$qp_STAR___52680.invoke(reducible.clj:92)"
  "query_processor.reducible$sync_qp$qp_STAR___52691.doInvoke(reducible.clj:127)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:362)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:358)"
  "query_processor$fn__63277$process_query_and_save_execution_BANG___63286$fn__63289.invoke(query_processor.clj:373)"
  "query_processor$fn__63277$process_query_and_save_execution_BANG___63286.invoke(query_processor.clj:366)"
  "query_processor$fn__63322$process_query_and_save_with_max_results_constraints_BANG___63331$fn__63334.invoke(query_processor.clj:385)"
  "query_processor$fn__63322$process_query_and_save_with_max_results_constraints_BANG___63331.invoke(query_processor.clj:378)"
  "api.dataset$run_query_async$fn__80098.invoke(dataset.clj:73)"
  "query_processor.streaming$streaming_response_STAR_$fn__48763$fn__48764.invoke(streaming.clj:164)"
  "query_processor.streaming$streaming_response_STAR_$fn__48763.invoke(streaming.clj:163)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:67)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:65)"
  "async.streaming_response$do_f_async$task__30404.invoke(streaming_response.clj:86)"],
 :card_id nil,
 :context :ad-hoc,
 :error "You do not have permissions to run this query.",
 :row_count 0,
 :running_time 0,
 :preprocessed
 {:database 1,
  :query
  {:source-table 7,
   :fields
   [[:field 59 nil]
    [:field 62 nil]
    [:field 64 nil]
    [:field 65 nil]
    [:field 50 nil]
    [:field 52 nil]
    [:field 53 nil]
    [:field 58 {:temporal-unit :default}]
    [:field 55 {:temporal-unit :default}]
    [:field 57 {:temporal-unit :default}]
    [:field 56 nil]
    [:field 63 nil]
    [:field 61 nil]
    [:field 60 nil]
    [:field 51 nil]
    [:field 54 nil]],
   :limit 1048575,
   :metabase.query-processor.middleware.limit/original-limit nil},
  :type :query,
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true},
  :info {:executed-by 2, :context :ad-hoc}},
 :ex-data
 {:type :missing-required-permissions,
  :required-permissions #{"/db/1/schema/PUBLIC/table/7/query/"},
  :actual-permissions
  #{"/query/db/1/schema/PUBLIC/table/8/"
    "/db/1/schema/PUBLIC/table/6/"
    "/collection/2/"
    "/collection/namespace/snippets/root/"
    "/db/1/schema/PUBLIC/table/5/"
    "/db/1/schema/PUBLIC/table/4/"
    "/db/1/schema/PUBLIC/table/2/"
    "/download/db/1/native/"
    "/collection/root/"
    "/data/db/1/schema/PUBLIC/table/8/"
    "/db/1/schema/PUBLIC/table/1/"
    "/download/db/1/"
    "/application/subscription/"
    "/db/1/schema/PUBLIC/table/8/"
    "/db/1/schema/PUBLIC/table/3/"},
  :permissions-error? true},
 :data {:rows [], :cols []}}
```
</details>

**To Reproduce**
1) create a user
2) In Admin settings go to Permissions - Databases, for Sample Database set permissions for AllUsers to Granular
3) Set Data Access for the table Accounts to "No self-service", Save changes
4) Create a simple question for Accounts table, add it to a dashboard
5) With a non Admin user, open the dashboard, from there open the question
6) Error message is displayed: You do not have permissions to run this query.

Note: It works when opening the dahsboard directly and it also works when opening the question directly!

**Expected behavior**
If you can see the card, you should be able to see the question. You shouldn't be able to do any further operations like drill down/through

**Screenshots/videos**
NA

**Information about your Metabase Installation:**
- Metabase version: 1.45.3/46-RC2

**Severity**
P1

**Additional context**
This bug also leads to:
1) a user who has collection permissions on the dashboard where the question is cannot add or edit filters in the dashboard
image

### PR DESCRIPTION
**Pull request by ranquild**
Fixes https://github.com/metabase/metabase/issues/29076
Reproduces https://github.com/metabase/metabase/issues/29076

The previous check for data permissions worked only when there is no sandboxing applied. With sandboxing, the `database` object is returned by api endpoints but the table might be not. So we change the check from `readOnly` to `isEditable` which also checks for metadata being present, which won't be the same if the user doesn't have self-service access. The query builder also relies on the `isEditable()` check ([example](https://github.com/metabase/metabase/blob/d03cc876e65e3c8215189aa3b92af3dcd5290b3c/frontend/src/metabase/query_builder/components/view/QuestionFilters.jsx#L177)), so it seems like the way forward.

The caveat here is that the change in the permission check affects click actions as well. To make the check work, we need to load metadata for cards referenced in click actions. Most of the PR is dealing with that. If we don't load metadata here `isEditable()` would always return false for the cards outside of the dashboard, like ones added via `Custom Destination -> Question`.

Screenshot 2023-04-03 at 18 47 22

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/29758)

### CODE REVIEW

**Review by kulyk — APPROVED** [review]

File: frontend/src/metabase/lib/click-behavior.js [hunk 1]
```diff
@@ -46,3 +48,28 @@ export function hasActionsMenu(dashcard) {
 export function isTableDisplay(dashcard) {
   return dashcard?.card?.display === "table";
 }
+
+export function getLinkTargets(settings) {
```
Comment (ranquild) [inline frontend/src/metabase/lib/click-behavior.js:52]: Moved these functions from `WithVizSettingsData`, they are not changed

File: frontend/src/metabase/dashboard/hoc/WithVizSettingsData.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import React from "react";
+import { connect } from "react-redux";
+import { withRouter } from "react-router";
+import _ from "underscore";
+
+import { getUserAttributes } from "metabase/selectors/user";
+import { getLinkTargets } from "metabase/lib/click-behavior";
+
+// This HOC give access to data referenced in viz settings.
+const WithVizSettingsData = ComposedComponent => {
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/hoc/WithVizSettingsData.js:11]: Moved from `visualizations` as it's only used in dashboards. The "composed" part of this component hasn't changed, but I've moved data loading for click behaviors to redux

File: frontend/src/metabase/dashboard/actions/metadata.js [hunk 3]
```diff
@@ -1,16 +1,24 @@
...
 import { isVirtualDashCard } from "../utils";
 
-export const loadMetadataForDashboard = dashCards => (dispatch, getState) => {
+export const loadMetadataForDashboard = dashCards => dispatch => {
+  const cards = dashCards
+    .filter(dc => !isVirtualDashCard(dc)) // exclude text cards
+    .flatMap(dc => [dc.card].concat(dc.series || []));
+
+  dispatch(loadMetadataForCards(cards));
+  dispatch(loadMetadataForLinkedTargets(dashCards));
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/metadata.js:14]: We need to load metadata for cards not only that are visible in the dashboard, but also referenced via click actions.

File: frontend/src/metabase/dashboard/actions/metadata.js [hunk 4]
```diff
@@ -1,16 +1,26 @@
...
 import { isVirtualDashCard } from "../utils";
 
-export const loadMetadataForDashboard = dashCards => (dispatch, getState) => {
+export const loadMetadataForDashboard = dashCards => async dispatch => {
+  const cards = dashCards
+    .filter(dc => !isVirtualDashCard(dc)) // exclude text cards
+    .flatMap(dc => [dc.card].concat(dc.series || []));
+
+  await Promise.all([
+    dispatch(loadMetadataForCards(cards)),
```
Comment (ranquild) [inline frontend/src/metabase/dashboard/actions/metadata.js:14]: Extracted loading metadata for cards in a function so I can call it for cards from click actions

**Comment by ranquild** [comment]
@kulyk the PR has changed significantly, could you take another look?

**Review by kulyk — APPROVED** [review]

**Pull request merged**